radix-engine-interface/src/blueprints/utils.rs

```diff
@@ -0,0 +1,99 @@
+use crate::blueprints::access_controller::*;
+use crate::blueprints::account::*;
+use crate::blueprints::identity::*;
+use crate::blueprints::package::*;
+use crate::blueprints::pool::*;
+use crate::internal_prelude::*;
+
+/// This function resolves the [`BlueprintId`] based on the entity type for native blueprints.
+///
+/// If the entity type is [`EntityType::GlobalGenericComponent`] then this function returns [`None`]
+/// since it can't tell what the [`BlueprintId`] of a generic component is. Otherwise, in the case
+/// where the [`EntityType`] belongs to a native blueprint this function will translate it into the
+/// appropriate [`BlueprintId`].
+pub fn resolve_main_module_blueprint_id(entity_type: EntityType) -> Option<&'static BlueprintId> {
+    match entity_type {
+        EntityType::GlobalPackage => Some(package_blueprint_id()),
+        EntityType::GlobalConsensusManager => Some(consensus_manager_blueprint_id()),
+        EntityType::GlobalValidator => Some(validator_blueprint_id()),
+        EntityType::GlobalTransactionTracker => Some(transaction_tracker_blueprint_id()),
+        EntityType::GlobalAccessController => Some(access_controller_blueprint_id()),
+        EntityType::GlobalOneResourcePool => Some(one_resource_pool_blueprint_id()),
+        EntityType::GlobalTwoResourcePool => Some(two_resource_pool_blueprint_id()),
+        EntityType::GlobalMultiResourcePool => Some(multi_resource_pool_blueprint_id()),
+        EntityType::GlobalAccountLocker => Some(account_locker_blueprint_id()),
+        EntityType::GlobalFungibleResourceManager => Some(fungible_resource_manager_blueprint_id()),
+        EntityType::GlobalNonFungibleResourceManager => {
+            Some(non_fungible_resource_manager_blueprint_id())
+        }
+        EntityType::InternalFungibleVault => Some(fungible_vault_blueprint_id()),
+        EntityType::InternalNonFungibleVault => Some(non_fungible_vault_blueprint_id()),
+        EntityType::GlobalAccount
+        | EntityType::GlobalPreallocatedSecp256k1Account
+        | EntityType::GlobalPreallocatedEd25519Account => Some(account_blueprint_id()),
+        EntityType::GlobalIdentity
+        | EntityType::GlobalPreallocatedSecp256k1Identity
+        | EntityType::GlobalPreallocatedEd25519Identity => Some(identity_blueprint_id()),
+        EntityType::InternalGenericComponent
+        | EntityType::InternalKeyValueStore
+        | EntityType::GlobalGenericComponent => None,
+    }
+}
+
+/// This function resolves the [`BlueprintId`] of the invoked blueprint given the [`BlueprintId`] of
+/// the main module and the [`ModuleId`] being invoked.
+pub fn resolve_invoked_blueprint_id(
+    main_module_blueprint_id: &BlueprintId,
+    module_id: ModuleId,
+) -> Option<&BlueprintId> {
+    match module_id {
+        ModuleId::Main => Some(main_module_blueprint_id),
+        // TODO: We could improve this to take into account if the blueprint has these modules or
```

Comment:
I agree it's not a super high priority, but I'd suggest we could move this to the `NativeInvocation` code, where we could have boolean flags on blueprint definitions saying e.g. `supports_metadata_module` etc?

Reply:
Potentially but it feels like mixing two concepts together: the resolution of the invoked blueprint id from the main module blueprint id and the module id and the typed invocation stuff. I view these two things as being completely separate.

Reply:
Fair enough - if we introduce a `NativeBlueprintDefinition` trait this is the kind of thing we can put on it I guess :D